checker/jdk/nullness/src/java/lang/ref/SoftReference.java

```diff
@@ -82,7 +83,7 @@
      *
      * @param referent object the new soft reference will refer to
      */
-    public SoftReference(T referent) {
+    public SoftReference(@Nullable T referent) {
```

Comment:
I expected to have no real opinion here, but after looking into it a little, I'll advocate further for `@Nullable`.

I don't personally need the constructor to accept `null`. I was just annotating Guava's subclasses ([example](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/base/FinalizableSoftReference.html)), and I noticed that at least our current version of the JDK supports `null`, so I put the annotation on the Guava and JDK classes.

- I am shocked to discover that ~7% of all `Reference` construction sites in the wider Google codebase pass a literal `null` argument. These mostly appear to be cases in which there will eventually be a "real" `Reference` but the code wants to eagerly initialize a field with _some_ `Reference` object to avoid leaving the field itself `null`. Grepping through the JDK sources shows some instances of the pattern there, too.

- Possibly some other number of `Reference` objects in the Google codebase are constructed with a non-constant value that _might_ be `null`. I doubt this is very common, but it's harder to easily search for. Of course, such users could always perform an explicit null check -- assuming that they don't need to end up with a `Reference` instance to initialize a field or something, as in the previous bullet.

- Here is some additional evidence that the JDK owners at least once upon a time considered null referents to be valid: https://bugs.openjdk.java.net/browse/JDK-4247236 As you note, though, there is no documentation of this that I can see.

Setting that aside for a moment: If you do end up wanting to forbid `null`, then we should probably change the type parameter to `<T extends [@NonNull Object>` so that users can't create a `SoftReference<@Nullable Object>` and pass in a null referent that way.

Reply:
> If you do end up wanting to forbid `null`, then we should probably change the type parameter to `<T extends [@NonNull Object>` so that users can't create a `SoftReference<@Nullable Object>` and pass in a null referent that way.

(Or we can just annotate the constructor parameter as `@NonNull`. I always forget about that :))

Reply:
OK, we can permit `null` as an argument to reference constructors.  I would personally prefer
 * use null instead of a nonsense reference, or
 * create a special nonsense reference, suppress the warning once at its construction, and use it where needed

However, I will bow to actual usage, including in the JDK.
Thanks for providing this extra evidence!